experimental/PropertyDDS/packages/property-shared-tree-interop/src/schemaConverter.ts

```diff
@@ -0,0 +1,329 @@
+	} else if (typeid === booleanType) {
+		valueSchema = ValueSchema.Boolean;
+	} else if (numberTypes.has(typeid) || isEnum) {
+		valueSchema = ValueSchema.Number;
+	} else {
+		// If this case occurs, there is definetely a problem with the ajv template,
+		// as unknown primitives should be issued there otherwise.
+		fail(`Unknown primitive typeid "${typeid}"`);
+	}
+	const treeSchema = builder.primitive(typeid, valueSchema);
+	treeSchemaMap.set(typeid, treeSchema);
+	return treeSchema;
+}
+
+function buildFieldSchema<Kind extends FieldKindTypes = FieldKindTypes>(
+	builder: SchemaBuilder,
+	treeSchemaMap: Map<string, LazyTreeSchema>,
+	allChildrenByType: InheritingChildrenByType,
+	fieldKind: Kind,
+	...fieldTypes: readonly string[]
+): FieldSchema {
+	if (fieldTypes.length === 0 || fieldTypes.find((t) => t === Any)) {
+		return SchemaBuilder.field(fieldKind, Any);
+	}
+	const allowedTypes = mapTypesAndChildren(
+		allChildrenByType,
+		(child) => buildTreeSchema(builder, treeSchemaMap, allChildrenByType, child),
+		...fieldTypes,
+	);
+	return SchemaBuilder.field(fieldKind, ...allowedTypes);
+}
+
+/**
+ * Creates a TypedSchemaCollection out of PropertyDDS schema templates.
+ * The templates must be registered beforehand using {@link PropertyFactory.register}.
+ * @param rootFieldKind - The kind of the root field.
+ * @param allowedRootTypes - The types of children nodes allowed for the root field.
+ * @param extraTypes - The extra types which can't be found when traversing across
+ * the PropertyDDS schema inheritances / dependencies starting from
+ * the root schema or built-in node property schemas.
+ */
+export function convertPropertyToSharedTreeSchema<Kind extends FieldKindTypes = FieldKindTypes>(
+	rootFieldKind: Kind,
+	allowedRootTypes: Any | ReadonlySet<string>,
+	extraTypes?: ReadonlySet<string>,
+) {
+	const builder = new SchemaBuilder("PropertyDDS to SharedTree schema builder");
+	const allChildrenByType = getAllInheritingChildrenTypes();
+	const treeSchemaMap: Map<string, LazyTreeSchema> = new Map();
+
+	const referencedTypeIDs =
+		allowedRootTypes === Any || allowedRootTypes.has(Any)
+			? new Set<string>()
+			: mapTypesAndChildren(allChildrenByType, (t) => t, ...allowedRootTypes);
+
+	primitiveTypes.forEach((primitiveType) => referencedTypeIDs.add(primitiveType));
+	// That's enough to add just "NodeProperty" type, as all other
+	// related built-in types will be added through inheritances.
+	referencedTypeIDs.add(nodePropertyType);
+
+	if (extraTypes) {
+		extraTypes.forEach((typeid) => referencedTypeIDs.add(typeid));
+	}
+
+	for (const referencedTypeId of referencedTypeIDs) {
+		buildTreeSchema(builder, treeSchemaMap, allChildrenByType, referencedTypeId);
+	}
+
+	const allowedTypes = allowedRootTypes === Any ? [Any] : [...allowedRootTypes];
+	const rootSchema = buildFieldSchema(
+		builder,
+		treeSchemaMap,
+		allChildrenByType,
+		rootFieldKind,
+		...allowedTypes,
+	);
+	return builder.intoDocumentSchema(rootSchema);
```

Comment:
A general way of using the schema remains same: global `treeSchema` map contains all node schema objects (not functions), and using `field.type` provides all type names.

Reply:
II am debugging the result of following

`const fullSchemaData1 = convertPropertyToSharedTreeSchema(FieldKinds.optional, allowedRoot);`

It looks like at treeSchema allowedTypes are really nodes, not functions.
At globalFieldSchema.rootFieldKey.schema.alloweedTypes there are still functions as follows:  () => cache.treeSchema

Reply:
Yes, `allowedTypes` by field schemas are not "normalized". Since the type-correct helpers for normalization (I mean, e.g., [`normalizeFlexListEager`](https://github.com/microsoft/FluidFramework/blob/38e62903a04919bb149564d28b36f6213d7ad8b1/experimental/dds/tree2/src/feature-libraries/modular-schema/typedSchema/flexList.ts#L24C17-L24C39) ) are not exposed into the public API of the SharedTree, I suppose, `allowedTypes` property is currently not intended for external use. We can ask MS guys. Currently, as I said, the way to use fields is to read their `types` and to lookup in the global treeSchema (same as it was also with an old schema API).